jupyter_console/ptshell.py

```diff
@@ -463,6 +472,16 @@ def pre_prompt(self):
             b.cursor_position += b.document.get_end_of_document_position()
 
     def interact(self, display_banner=None):
+        # Start handle_iopub thread for external input
+        def external_monitor():
+            while self.keep_running:
+                self.handle_iopub()
```

Comment:
We need to be careful about threading because sockets aren't threadsafe. Putting `handle_iopub` here means that `handle_iopub` must never be called from the main thread while this thread is running.